awssqs/pkg/reconciler/awssqssource.go

```diff
@@ -18,148 +18,89 @@ package reconciler
-	eventTypeReconciler eventtype.Reconciler
-
+// Reconciler implements controller.Reconciler for AwsSqsSource resources.
+type Reconciler struct {
 	receiveAdapterImage string
-}
-
-func (r *reconciler) InjectClient(c client.Client) error {
-	r.client = c
-	r.eventTypeReconciler.Client = c
-	return nil
-}
 
-func (r *reconciler) InjectConfig(c *rest.Config) error {
-	var err error
-	r.dynamicClient, err = dynamic.NewForConfig(c)
-	return err
+	kubeClientSet     kubernetes.Interface
+	eventingClientSet eventingclientset.Interface
+	eventTypeLister   eventinglisters.EventTypeLister
+	sinkResolver      *resolver.URIResolver
 }
 
-func (r *reconciler) Reconcile(ctx context.Context, object runtime.Object) error {
-	logger := logging.FromContext(ctx).Desugar()
+// Check that our Reconciler implements Interface
+var _ awssqssource.Interface = (*Reconciler)(nil)
 
-	var err error
-
-	src, ok := object.(*v1alpha1.AwsSqsSource)
-	if !ok {
-		logger.Error("could not find AwsSqs source", zap.Any("object", object))
-		return nil
-	}
+// ReconcileKind implements Interface.ReconcileKind.
+func (r *Reconciler) ReconcileKind(ctx context.Context, src *v1alpha1.AwsSqsSource) pkgreconciler.Event {
+	src.Status.InitializeConditions()
+	src.Status.ObservedGeneration = src.Generation
 
 	// This Source attempts to reconcile three things.
 	// 1. Determine the sink's URI.
 	//     - Nothing to delete.
 	// 2. Create a receive adapter in the form of a Deployment.
 	//     - Will be garbage collected by K8s when this AwsSqsSource is deleted.
 	// 3. Register that receive adapter as a Pull endpoint for the specified AWS SQS queue.
-	// 4. Create the EventTypes that it can emit.
-	//     - Will be garbage collected by K8s when this AwsSqsSource is deleted.
-
-	// See if the source has been deleted.
-	deletionTimestamp := src.DeletionTimestamp
-	if deletionTimestamp != nil {
-		r.removeFinalizer(src)
-		return nil
-	}
 
-	r.addFinalizer(src)
-
-	src.Status.ObservedGeneration = src.Generation
-	src.Status.InitializeConditions()
-
-	sinkURI, err := sinks.GetSinkURI(ctx, r.client, src.Spec.Sink, src.Namespace)
-	if err != nil {
+	if src.Spec.Sink != nil {
+		dest := &duckv1.Destination{
+			Ref: &duckv1.KReference{
+				Kind:       src.Spec.Sink.Kind,
+				Namespace:  src.Namespace,
+				Name:       src.Spec.Sink.Name,
+				APIVersion: src.Spec.Sink.APIVersion,
+			},
+		}
+		sinkURI, err := r.sinkResolver.URIFromDestinationV1(*dest, src)
+		if err != nil {
+			src.Status.MarkNoSink("NotFound", "")
+			return err
+		}
+		src.Status.MarkSink(sinkURI.String())
+	} else {
 		src.Status.MarkNoSink("NotFound", "")
-		return err
+		return nil
 	}
-	src.Status.MarkSink(sinkURI)
 
-	_, err = r.createReceiveAdapter(ctx, src, sinkURI)
-	if err != nil {
-		logger.Error("Unable to create the receive adapter", zap.Error(err))
+	if _, err := r.createReceiveAdapter(ctx, src); err != nil {
+		logging.FromContext(ctx).Error("Unable to create the receive adapter", zap.Error(err))
 		return err
```

Comment:
In other genreconciler based reconcilers, we seem to also have a `newDeploymentFailed()` function returned which also passes the error, should we also use that here?  Shouldn't we also call `src.Status.MarkNotDeployed()`?

Reply:
This is now what this did originally. There is lots to improve in this controller. That is not my aim in the PR. Just moving what was there to `+genreconciler`